tftp/util.py

```diff
@@ -2,126 +2,73 @@
+from itertools import tee
 from twisted.internet import reactor
-from twisted.internet.defer import maybeDeferred
+from twisted.internet.defer import CancelledError, maybeDeferred, succeed
+from twisted.internet.task import deferLater
 
 
-__all__ = ['SequentialCall', 'Spent', 'Cancelled', 'deferred']
+__all__ = ['CANCELLED', 'deferred', 'timedCaller']
 
 
-class Spent(Exception):
-    """Trying to iterate a L{SequentialCall}, that is exhausted"""
+# Token used by L{timedCaller} to denote that it was cancelled instead of
+# finishing naturally. This is not an error condition, and may not be of
+# interest during normal use, but is helpful when testing.
+CANCELLED = "CANCELLED"
 
-class Cancelled(Exception):
-    """Trying to iterate a L{SequentialCall}, that's been cancelled"""
 
-def no_op(*args, **kwargs):
-    pass
+def timedCaller(timings, call, last, clock=reactor):
+    """Call C{call} or C{last} according to C{timings}.
 
-class SequentialCall(object):
-    """Calls a given callable at intervals, specified by the L{timeout} iterable.
-    Optionally calls a timeout handler, if provided, when there are no more timeout values.
+    The given C{timings} is an iterable of numbers. Each is a delay in seconds
+    that will be taken before making the next call to C{call} or C{last}.
 
-    @param timeout: an iterable, that yields valid _seconds arguments to
-    L{callLater<twisted.internet.interfaces.IReactorTime.callLater>} (floats)
-    @type timeout: any iterable
+    The call to C{last} will happen after the last delay. If C{timings} is an
+    infinite iterable then C{last} will never be called.
 
-    @param run_now: whether or not the callable should be called immediately
-    upon initialization. Relinquishes control to the reactor
-    (calls callLater(0,...)). Default: C{False}.
-    @type run_now: C{bool}
-
-    @param callable: the callable, that will be called at the specified intervals
-    @param callable_args: the arguments to call it with
-    @param callable_kwargs: the keyword arguments to call it with
-
-    @param on_timeout: the callable, that will be called when there are no more
-    timeout values
-    @param on_timeout_args: the arguments to call it with
-    @param on_timeout_kwargs: the keyword arguments to call it with 
-    
+    This returns a C{Deferred} which can be cancelled. If the cancellation is
+    successful — i.e. there is something to cancel — then the result is set to
+    L{CANCELLED}. In other words, L{CancelledError} is squashed into a
+    non-failure condition.
     """
+    timings = iterlast(timings)
+
+    def iterate(_=None):
+        for is_last, delay in timings:
+            # Return immediately with a deferred call.
+            if is_last:
+                return deferLater(clock, delay, last)
+            else:
+                return deferLater(clock, delay, call).addCallback(iterate)
+        else:
+            # No timings were given.
+            return succeed(None)
```

Comment:
Feel like either no timings should raise an error or should just call `last` instantly (but that is some what of a weird behavior).

Strange that you have it just resolving to None.